services/requester/cross-spork_client.go

```diff
@@ -2,16 +2,87 @@ package requester
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"github.com/onflow/cadence"
 	"github.com/onflow/flow-go-sdk"
 	"github.com/onflow/flow-go-sdk/access"
 	"github.com/rs/zerolog"
-	"golang.org/x/exp/maps"
 	"golang.org/x/exp/slices"
 )
 
+var ErrOutOfRange = errors.New("height is out of range for provided spork clients")
+
+type sporkClient struct {
+	firstHeight uint64
+	lastHeight  uint64
+	client      access.Client
+}
+
+// contains checks if the provided height is withing the range of available heights
+func (s *sporkClient) contains(height uint64) bool {
+	return height >= s.firstHeight && height <= s.lastHeight
+}
+
+type sporkClients []*sporkClient
+
+// addSpork will add a new spork host defined by the first and last height boundary in that spork.
+func (s *sporkClients) add(client access.Client) error {
+	header, err := client.GetLatestBlockHeader(context.Background(), true)
+	if err != nil {
+		return fmt.Errorf("could not get latest height using the spork client: %w", err)
+	}
+
+	info, err := client.GetNodeVersionInfo(context.Background())
+	if err != nil {
+		return fmt.Errorf("could not get node info using the spork client: %w", err)
+	}
+
+	*s = append(*s, &sporkClient{
+		firstHeight: info.NodeRootBlockHeight,
+		lastHeight:  header.Height,
+		client:      client,
+	})
+
+	return nil
+}
+
+// get spork client that contains the height or nil if not found.
+func (s *sporkClients) get(height uint64) access.Client {
+	for _, spork := range *s {
+		if spork.contains(height) {
+			return spork.client
+		}
+	}
+
+	return nil
+}
+
+// continuous checks if all the past spork clients create a continuous
+// range of heights.
+func (s *sporkClients) continuous() bool {
+	firsts := make([]uint64, len(*s))
+	lasts := make([]uint64, len(*s))
+
+	for i, c := range *s {
```

Comment:
nit: I think you could simplify this logic by sorting the list of clients first, then iterating over the list once.

Reply:
Good idea, added.